lib/termination-poller.js

```diff
@@ -0,0 +1,141 @@
+const assert = require('assert');
+const {runAWSRequest} = require('./aws-request');
+const Iterate = require('taskcluster-lib-iterate');
+const log = require('./log');
+
+const INSTANCE_LIMIT = 200;
+const POLL_DELAY = 30;
+
+class TerminationPoller {
+  constructor({ec2, regions, runaws = runAWSRequest, pollDelay = POLL_DELAY, state}) {
+    assert(typeof ec2 === 'object');
+    assert(Array.isArray(regions));
+    regions.forEach(region => assert(typeof region === 'string'));
+    assert(typeof state === 'object');
+
+    this.ec2 = ec2;
+    this.regions = regions;
+    this.state = state;
+    this.runaws = runaws;
+
+    this.iterator = new Iterate({
+      maxIterationTime: 60 * 15,
+      watchDog: 60 * 15,
+      maxFailures: 10,
+      waitTime: pollDelay,
+      handler: async(watchdog) => {
+        try {
+          await this.poll(watchdog);
+        } catch (err) {
+          log.error({err}, 'iteration error');
+        }
+      },
+    });
+  }
+
+  start() {
+    return this.iterator.start();
+  }
+
+  stop() {
+    return this.iterator.stop();
+  }
+
+  async poll(watchdog, sharedState) {
+    let toPoll = await this.state.findTerminationsToPoll(INSTANCE_LIMIT);
+
+    if (toPoll.length && toPoll.length === 0) {
```

Comment:
yep.  an even stronger check would be `typeof topoll.length === 'number'`, but since this function is one that we control, I'll just ensure that it's always going to return a list